### PR TITLE
hwmonitor@sylfurd : Applet crashed in Cinnamon < 4.x

### LINKED ISSUE #2393: [hwmonitor@sylfurd]: Can not add to panel
**Facts**
 * Applet version/Build date (to be found in "About.." or in "metadata.json")
 N/A (git rev https://github.com/linuxmint/cinnamon-spices-applets/commit/714987a176264c812b3cd41f72b045e0df3236d1 I believe)
 * Cinnamon version (cinnamon --version)
  Cinnamon 3.2.7
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...)
  Debian GNU/Linux 9.8 (stretch)

**Notify author of applet**
@sylfurd (not listed in [CODEOWENERS](https://github.com/linuxmint/cinnamon-spices-applets/blob/master/.github/CODEOWNERS)

**Issue**

The applet can not be added to the panel.

  There was a problem loading the selected item, and it has been disabled.

`glass.log` looks like this:
```
error t=2019-03-29T18:05:57.133Z /org/freedesktop/UPower/devices/battery_BAT0:=LGN
error t=2019-03-29T18:06:30.758Z Array(...).fill is not a function
trace t=2019-03-29T18:06:30.758Z 
<----------------
Graph.prototype._init@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:154
Graph@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:146
GraphicalHWMonitorApplet.prototype._init@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:91
GraphicalHWMonitorApplet@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:58
main@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:310
createApplet@/usr/share/cinnamon/js/ui/appletManager.js:404
addAppletToPanels@/usr/share/cinnamon/js/ui/appletManager.js:294
finishExtensionLoad@/usr/share/cinnamon/js/ui/appletManager.js:59
loadExtension@/usr/share/cinnamon/js/ui/extension.js:451
onEnabledAppletsChanged@/usr/share/cinnamon/js/ui/appletManager.js:235
---------------->
error t=2019-03-29T18:06:30.758Z [Applet "hwmonitor@sylfurd"]: Failed to evaluate 'main' function on applet: hwmonitor@sylfurd/64
error t=2019-03-29T18:06:30.758Z Could not load applet hwmonitor@sylfurd
```

Thanks for looking into this.

### PR DESCRIPTION
**Pull request by Hultan**
The applet crashed in Cinnamon 3.x because of some newer javascript syntax.

Closes #2393

### CODE REVIEW

**Comment by jkirk** [comment]
Just tested the PR. Unfortunately I get:
```
error t=2019-10-14T19:59:24.072Z missing : after property id
trace t=2019-10-14T19:59:24.072Z■
<----------------
@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:42
Extension.prototype._init@/usr/share/cinnamon/js/ui/extension.js:155
Extension@/usr/share/cinnamon/js/ui/extension.js:116
loadExtension@/usr/share/cinnamon/js/ui/extension.js:449
onEnabledAppletsChanged@/usr/share/cinnamon/js/ui/appletManager.js:235
---------------->
error t=2019-10-14T19:59:24.072Z [Applet "hwmonitor@sylfurd"]: Error importing applet.js from hwmonitor@sylfurd
```

File: hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js [hunk 1]
```diff
@@ -4,8 +4,12 @@ const Main = imports.ui.main;
...
 // Class responsible for drawing an instance of one graph
 //
-class Graph {    
-    constructor (provider, graphArea, theme, show_detail) {
+function Graph(provider, graphArea, theme, show_detail) {
+    this.init(provider, graphArea, theme, show_detail);
+}
+
+Graph.prototype = {    
+    init(provider, graphArea, theme, show_detail) {
```
Comment (jkirk) [inline hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js]: It looks like, it needs to be something like this:
```
    init: function(provider, graphArea, theme, show_detail) {
```

File: hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js [hunk 2]
```diff
@@ -16,7 +20,7 @@ class Graph {
         for (let i = 0; i < this.datas.length; i++) {
             this.datas[i] = 0;
         }
-    }
+    },
 
     paint(area) {
```
Comment (jkirk) [inline hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js]: Same here:
```
    paint: function(area) {
```

File: hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js [hunk 3]
```diff
@@ -48,7 +52,7 @@ class Graph {
 
         // Reset translation
         cr.translate( -(this.graph.outer.left + 0.5), -(this.graph.outer.top + 0.5));
-    }
+    },
 
     // Draw the border around the graph
     drawBorder(cr){
```
Comment (jkirk) [inline hwmonitor@sylfurd/files/hwmonitor@sylfurd/graph.js]: And so on...

File: hwmonitor@sylfurd/files/hwmonitor@sylfurd/providers.js [hunk 4]
```diff
@@ -21,8 +21,12 @@ try {
...
 
 // Class responsible for getting CPU data
-class CpuDataProvider {
-	constructor() {
+function CpuDataProvider() {
+    this.init();
+}
+
+CpuDataProvider.prototype = {
+	init() {
```
Comment (jkirk) [inline hwmonitor@sylfurd/files/hwmonitor@sylfurd/providers.js]: Same here:
```
    init: function() {
```

File: hwmonitor@sylfurd/files/hwmonitor@sylfurd/providers.js [hunk 5]
```diff
@@ -31,7 +35,7 @@ class CpuDataProvider {
         this.last_delta = 0;
         this.name = _("CPU");
         this.type = "CPU";
-    }
+    },
 
     getData() {
```
Comment (jkirk) [inline hwmonitor@sylfurd/files/hwmonitor@sylfurd/providers.js]: And so on...

**Comment by Hultan** [comment]
First of all, why didn't my machine complain? Weird :-)

Anyway, fixed it. Hopefully, it works this time...

**Comment by jkirk** [comment]
Yes, weird :-) Maybe the new JavaScript engine is more forgiving (aka supports the new syntax)?

"Better" now, but still some issues:
```
info t=2019-10-14T21:51:07.350Z Installing settings for hwmonitor@sylfurd
error t=2019-10-14T21:51:07.351Z Settings key label_size is missing property units
error t=2019-10-14T21:51:07.351Z Could not set up binding - settings object was not initialized successfully for hwmonitor@sylfurd
[...]
error t=2019-10-14T21:51:07.354Z Could not set up binding - settings object was not initialized successfully for hwmonitor@sylfurd
error t=2019-10-14T21:51:07.354Z rgb is undefined
trace t=2019-10-14T21:51:07.355Z
<----------------
GraphicalHWMonitorApplet.prototype.getColors@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:306
GraphicalHWMonitorApplet.prototype.createThemeObject@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:318
GraphicalHWMonitorApplet.prototype._init@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:142
GraphicalHWMonitorApplet@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:64
main@/home/jkirk/.local/share/cinnamon/applets/hwmonitor@sylfurd/applet.js:381
createApplet@/usr/share/cinnamon/js/ui/appletManager.js:404
addAppletToPanels@/usr/share/cinnamon/js/ui/appletManager.js:294
finishExtensionLoad@/usr/share/cinnamon/js/ui/appletManager.js:59
loadExtension@/usr/share/cinnamon/js/ui/extension.js:451
onEnabledAppletsChanged@/usr/share/cinnamon/js/ui/appletManager.js:235
---------------->
error t=2019-10-14T21:51:07.355Z [Applet "hwmonitor@sylfurd"]: Failed to evaluate 'main' function on applet: hwmonitor@sylfurd/158
error t=2019-10-14T21:51:07.355Z Could not load applet hwmonitor@sylfurd
```

**Comment by jkirk** [comment]
I think the problem is caused by the way how the settings of the applet are set up.
`this.settings.bind` (see:  [#L95](https://github.com/linuxmint/cinnamon-spices-applets/blob/4d70956a3befdc2da4d13fada897b03486b2cd4e/hwmonitor%40sylfurd/files/hwmonitor%40sylfurd/applet.js#L95)) vs `this.settings.bindProperty` in all other applets I have installed.

At first "just" errors `Could not set up binding - settings object was not initialized successfully for hwmonitor@sylfurd` are shown but as `border_color` in [#L98](https://github.com/linuxmint/cinnamon-spices-applets/blob/4d70956a3befdc2da4d13fada897b03486b2cd4e/hwmonitor%40sylfurd/files/hwmonitor%40sylfurd/applet.js#L98) is not loaded, the `createThemeObject` function fails when trying to call `getColors` in [#L318](https://github.com/linuxmint/cinnamon-spices-applets/blob/4d70956a3befdc2da4d13fada897b03486b2cd4e/hwmonitor%40sylfurd/files/hwmonitor%40sylfurd/applet.js#L318).

**Comment by Hultan** [comment]
@jkirk , this is getting more complicated than just remove the `class` keyword. I am going to have to set up a virtual machine with Cinnamon 3.2 on and really test it this time. This will probably take me a day or two, I might even need the weekend. So I will close this PR now and create a new one when I have tested it in Cinnamon 3.2.x...

**Comment by claudiux** [comment]
@Hultan, I suggest you to do a multi-version of this applet.

Kind regards.

**Comment by Hultan** [comment]
This is weird, I followed @claudiux direction and created a multiversion applet, and it still does not work in Cinnamon 3.2.7, even if I go back to this super old version (from November last year, before I started messing with this applet):

https://github.com/linuxmint/cinnamon-spices-applets/tree/714987a176264c812b3cd41f72b045e0df3236d1/hwmonitor%40sylfurd/files/hwmonitor%40sylfurd

I am getting the following glass.log:

`error t=2019-10-15T09:07:20.991Z missing } after function body
trace t=2019-10-15T09:07:20.991Z 
<----------------
Extension.prototype._init@/usr/share/cinnamon/js/ui/extension.js:155
Extension@/usr/share/cinnamon/js/ui/extension.js:116
loadExtension@/usr/share/cinnamon/js/ui/extension.js:449
onEnabledAppletsChanged@/usr/share/cinnamon/js/ui/appletManager.js:235
---------------->
error t=2019-10-15T09:07:20.991Z [Applet "hwmonitor@sylfurd"]: Error importing applet.js from hwmonitor@sylfurd
error t=2019-10-15T09:07:20.992Z Could not load applet hwmonitor@sylfurd
info t=2019-10-15T09:07:20.994Z Adding systray: mintupdate.py (24x24px)
info t=2019-10-15T09:07:20.995Z Hiding systray: network
info t=2019-10-15T09:07:21.068Z Adding systray: mintupdate.py (24x24px)
info t=2019-10-15T09:07:21.069Z Hiding systray: network
info t=2019-10-15T09:07:22.588Z Resized mintupdate.py with normalized size (22x22px)`

You can try it out from my fork, @jkirk , but it does not work on my Linux Mint 18.1 (Cinnamon 3.2.7):

https://github.com/SoftTeamAB/cinnamon-spices-applets

**Comment by Odyseus** [comment]
Hello, everybody.

@Hultan: Several errors in your fork:

- Do not use 2.8 as a version, use **3.2**. There isn't a currently supported distribution that uses Cinnamon version 2.x (Linux Mint 17.x has been EOL (End Of Life) for more than six months now).
- Copy **ALL** files into the **3.2** folder **EXCEPT** for the **icon.png** and **metadata.json** files. Do the same for the **3.8** folder but do not modify any of those files.
- Convert to *JavaScrpt prototypes* **ALL** .js files, not just the **applet.js** file.

**Comment by Hultan** [comment]
Well, now I have finally kind of gotten it to work. The 3.2-version is from last November though, I could not get the configurations to work with the latest version (with the new graphs and all the settings I have added lately). 

So not perfect, but now it should work in 3.2.x. Hopefully I will find some time soon to make the 3.2 version to be pretty much the same as the 3.8 version...

**Comment by claudiux** [comment]
@Hultan I will try to fix the code. I'm used to multi-version.

**Comment by Hultan** [comment]
If you do so, the problem wasn't with the code syntax, that was just to replace the classes with prototypes. The problem was with setting up the settings and creating the theme-object (createThemeObject()), possibly because some setting (border_color) was not initialized? My entire VM hanged when it loaded the applet. 

So I don't think the problem is with the multi-versioning, that part seems to work just fine after I renamed 2.8 to 3.2.

**Comment by claudiux** [comment]
There is a (first?) problem with `settings-schema.json`, which uses layout and pages, unknown before Cinnamon 3.8.

**Comment by Hultan** [comment]
Aha, that was in my last PR, I cleaned up the config dialog, the only thing I changed except the layout bit was that I split up the theme into two combos instead of one. I can fix that if you want? Otherwise it should just be a copy of the previous settings-schema.json and copy the two theme combos from the current settings-schema.json.

**Comment by claudiux** [comment]
@ Hultan I have edited my previous message. If you can do it now, do it. I have to leave a few hours.

**Comment by claudiux** [comment]
@Hultan 

`error t=2019-10-15T15:14:41.413Z Settings key label_size is missing property units`

Try that: in `3.2/settings-schema.json`, line 65, is the `label_size` key. Add a property "units". And so on for each key having type "spinbutton".

End of line 35, in the same file: a `)` is missing before the last `"`!!!

**Comment by Hultan** [comment]
Lol, now I feel stupid! I didn't even notice the underscore between the words label and size (label_size) so I never realized that it was one of my own settings! :-)

Thank you @claudiux, now it works!

One small thing though, I use `"dependency": "theme=user"` to control the visibility of certain settings depending on the value of a combobox (theme). This works in 3.8 but not in 3.2, the settings are always visible. Is that something new in 3.8 too?

**Comment by Hultan** [comment]
Ok, now finally we can merge this pull request. Got it to work decently in 3.2...

Closing #2393

**Comment by claudiux** [comment]
Congrats @Hultan !

And yes, "dependency" is declared as "New in Cinnamon 3.8".

**Comment by Hultan** [comment]
Just out of curiosity, where do you go to find out what's new in each version?

**Comment by claudiux** [comment]
`apt install devhelp gir1.2-devhelp-3.0 libgtk-3-doc libglib2.0-doc libpango1.0-doc cinnamon-doc` then use devhelp to go to Cinnamon Tutorials -> Extension system -> Applet, desklet and extension settings reference.

You can also download all (latest) versions of Cinnamon and look at the files in the `cinnamon-x.y.z/files/usr/share/cinnamon/applets/settings-example@cinnamon.org/` folder.

**Comment by Odyseus** [comment]
Hello, everybody.

@Hultan: In addition to the documentations suggested by **claudiux**, it also exists [this on-line documentation](https://lazka.github.io/pgi-docs/) about GObject libraries (Clutter, GLib, Gio, etc.). These libraries are greatly used by Cinnamon itself and by Cinnamon xlets. It can also be [downloaded for off-line use](https://github.com/lazka/pgi-docs) as a static HTML site, and also can be downloaded as [Devhelp books](https://github.com/pygobject/pgi-docs-devhelp).

**Pull request merged**